### PR TITLE
Add `length` F32 tests

### PR DESCRIPTION
**Pull request by zoddicus**
Implements support for Vector to F32 at various points in the testing
framework.

Refactors vector test case generation code out of `dot.spec.ts` into a
common code location.

Issue #1216

<hr>

**Requirements for PR author:**

- [x] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [x] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [x] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [ ] Tests are properly located in the test tree.
- [ ] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [ ] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

**Comment by zoddicus** [comment]
PTAL, this is ready for review. I have update the tests to reflect the changes to the spec required for scalars

File: src/webgpu/shader/execution/expression/call/builtin/length.spec.ts [hunk 1]
```diff
@@ -9,28 +9,81 @@ Returns the length of e (e.g. abs(e) if T is a scalar, or sqrt(e[0]^2 + e[1]^2 +
...
+    u.combine('inputSource', allInputSources)
   )
-  .unimplemented();
+  .fn(async t => {
+    const makeCase = (x: number): Case => {
+      return makeUnaryToF32IntervalCase(x, lengthInterval);
+    };
+    const cases: Case[] = fullF32Range().map(makeCase);
+
+    await run(t, builtin('length'), [TypeF32], TypeF32, t.params, cases);
```
Comment (ben-clayton) [inline src/webgpu/shader/execution/expression/call/builtin/length.spec.ts:52]: Did we decide that this needs to allow `x` and `x*x`?

File: src/webgpu/util/math.ts [hunk 2]
```diff
@@ -449,6 +449,35 @@ export function sparseF32Range(): Array<number> {
...
+ *
+ * The tests do not do the simple option for coverage of computing the cartesian
+ * product of all of the interesting float values N times for vecN tests,
+ * because that creates a huge number of tests for vec3 and vec4, leading to
+ * time outs.
+ * Instead they insert the interesting f32 values into each location of the
+ * vector to get a spread of testing over the entire range. This reduces the
+ * number of cases being run substantially, but maintains coverage.
+ */
+export const kVectorTestValues = {
```
Comment (ben-clayton) [inline src/webgpu/util/math.ts:463]: Maybe throw in some negatives?

File: src/unittests/f32_interval.spec.ts [hunk 3]
```diff
@@ -2435,6 +2436,55 @@ g.test('smoothStepInterval')
...
+interface VectorToIntervalCase {
+  input: number[];
+  expected: IntervalBounds;
+}
+
+g.test('lengthInterval')
+  .paramsSubcasesOnly<VectorToIntervalCase>(
+    // prettier-ignore
+    [
+      // Some of these are hard coded, since the error intervals are difficult to express in a closed human readable
```
Comment (ben-clayton) [inline src/unittests/f32_interval.spec.ts:2448]: What about scalars?

File: src/webgpu/shader/execution/expression/call/builtin/length.spec.ts (hunk 1 above)
Comment (zoddicus) [reply src/webgpu/shader/execution/expression/call/builtin/length.spec.ts:52]: Yup, sqrt(x*x) for scalars, https://github.com/gpuweb/gpuweb/pull/3409

File: src/webgpu/util/math.ts (hunk 2 above)
Comment (zoddicus) [reply src/webgpu/util/math.ts:463]: Done

File: src/unittests/f32_interval.spec.ts (hunk 3 above)
Comment (zoddicus) [reply src/unittests/f32_interval.spec.ts:2487]: Done

**Comment by zoddicus** [comment]
PTAL, I think this is ready for another round of review

File: src/webgpu/util/f32_interval.ts [hunk 4]
```diff
@@ -80,7 +81,8 @@ export class F32Interval {
 
   /** @returns a string representation for logging purposes */
   public toString(): string {
-    return `[${this.bounds()}]`;
+    // return `[${this.bounds()}]`;
```
Comment (ben-clayton) [inline src/webgpu/util/f32_interval.ts:84]: Remove?

**Review by ben-clayton — APPROVED** [review]

**Pull request merged**